lib/Twig/Extension/Core.php

```diff
@@ -652,11 +652,8 @@ function twig_array_merge($arr1, $arr2)
 function twig_slice(Twig_Environment $env, $item, $start, $length = null, $preserveKeys = false)
 {
     if ($item instanceof Traversable) {
-        if ($item instanceof IteratorAggregate) {
+        if ($start >= 0 && $length >= 0 && $item instanceof IteratorAggregate) {
             $item = $item->getIterator();
-        }
-
-        if ($start >= 0 && $length >= 0 && $item instanceof Iterator) {
```

Comment:
and you can also pass an Iterator directly, which would not be an IteratorAggregate. This would be the main thing you were breaking